src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -952,90 +969,98 @@ private struct CurrentState
+            /// <summary>Gets the preferred DFA state for nullability. In DFA mode this is just the state itself.</summary>
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            public static bool IsDeadend(ref CurrentState state) => state.DfaState!.IsDeadend;
+            public static int ExtractNullableCoreStateId(SymbolicRegexMatcher<TSet> matcher, ref CurrentState state, ReadOnlySpan<char> input, int pos) => state.DfaStateId;
 
             /// <summary>Gets the length of any fixed-length marker that exists for this state, or -1 if there is none.</summary>
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            public static int FixedLength(ref CurrentState state, uint nextCharKind) => state.DfaState!.FixedLength(nextCharKind);
-
-            /// <summary>Gets whether this is an initial state.</summary>
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            public static bool IsInitialState(ref CurrentState state) => state.DfaState!.IsInitialState;
+            public static int FixedLength(SymbolicRegexBuilder<TSet> builder, ref CurrentState state, uint nextCharKind) => state.DfaState(builder)!.FixedLength(nextCharKind);
 
             /// <summary>Take the transition to the next DFA state.</summary>
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             public static bool TakeTransition(SymbolicRegexBuilder<TSet> builder, ref CurrentState state, int mintermId)
             {
-                Debug.Assert(state.DfaState is not null, $"Expected non-null {nameof(state.DfaState)}.");
+                Debug.Assert(state.DfaStateId > 0, $"Expected non-zero {nameof(state.DfaStateId)}.");
                 Debug.Assert(state.NfaState is null, $"Expected null {nameof(state.NfaState)}.");
                 Debug.Assert(builder._delta is not null);
 
-                // Get the current state.
-                DfaMatchingState<TSet> dfaMatchingState = state.DfaState!;
-
                 // Use the mintermId for the character being read to look up which state to transition to.
                 // If that state has already been materialized, move to it, and we're done. If that state
                 // hasn't been materialized, try to create it; if we can, move to it, and we're done.
-                int dfaOffset = (dfaMatchingState.Id << builder._mintermsLog) | mintermId;
-                DfaMatchingState<TSet>? nextState = builder._delta[dfaOffset];
-                if (nextState is not null || builder.TryCreateNewTransition(dfaMatchingState, mintermId, dfaOffset, checkThreshold: true, out nextState))
+                int dfaOffset = (state.DfaStateId << builder._mintermsLog) | mintermId;
+                int nextStateId = builder._delta[dfaOffset];
+                if (nextStateId > 0 || builder.TryCreateNewTransition(state.DfaStateId, mintermId, dfaOffset, checkThreshold: true, out nextStateId))
```

Comment:
You're right, this is cleaner.

Reply:
Ah, now I remember why I added it: having a wrapper for the state ID allowed a single if body here to handle both an existing transition and a new one. I'll still remove the wrapper, since expanding the code here makes the different cases a bit more apparent anyway.